File: VB/WindowsApplication197/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Repository
Imports System.Reflection
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Filtering
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Filtering.Helpers

Namespace WindowsApplication197
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()

			InitData()
			gridControl1.DataSource = dv

			Dim columnCollection As New DevExpress.XtraGrid.FilterEditor.ViewFilterColumnCollection(gridView1)
			Dim ri As New RepositoryItemPopupContainerEdit()
			ri.PopupControl = popupContainerControl1
			ri.CloseOnOuterMouseClick = False
			AddHandler ri.QueryResultValue, AddressOf ri_QueryResultValue
			AddHandler ri.QueryPopUp, AddressOf ri_QueryPopUp
			columnCollection("Date").SetColumnEditor(ri)
			columnCollection("ID").SetColumnEditor(New RepositoryItemCalcEdit())
			filterControl1.SetFilterColumnsCollection(columnCollection)

		End Sub

		Private Sub ri_QueryPopUp(ByVal sender As Object, ByVal e As CancelEventArgs)
			dateEdit1.EditValue = (CType(sender, PopupContainerEdit)).EditValue
		End Sub

		Private Sub ri_QueryResultValue(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.QueryResultValueEventArgs)
			e.Value = dateEdit1.EditValue
			Dim [date] As DateTime = Convert.ToDateTime(e.Value)
			Dim pi As PropertyInfo = GetType(FilterControl).GetProperty("FocusInfo", BindingFlags.Instance Or BindingFlags.NonPublic)
			Dim focus As FilterControlFocusInfo = CType(pi.GetValue(filterControl1, Nothing), FilterControlFocusInfo)
			Dim cnode As ClauseNode = TryCast(focus.Node, ClauseNode)
			If cnode Is Nothing Then
				Return
			End If
			If cnode.Operation = ClauseType.Between Then
				If focus.ElementIndex = 2 Then
					PatchOperator(TryCast(cnode.AdditionalOperands(1), OperandValue), [date], comboBoxEdit1.Text, True)
				ElseIf focus.ElementIndex = 3 Then
					PatchOperator(TryCast(cnode.AdditionalOperands(0), OperandValue), [date], comboBoxEdit1.Text, False)
				End If
				cnode.RecalcLabelInfo()
			End If
		End Sub

		Private Sub PatchOperator(ByVal opr As OperandValue, ByVal [date] As DateTime, ByVal range As String, ByVal isFirst As Boolean)
			If (CObj(opr)) Is Nothing Then
				Return
			End If
			Select Case range
				Case "Day"
					If isFirst Then
						opr.Value = [date].AddDays(1)
					Else
						opr.Value = [date].AddDays(-1)
					End If
				Case "Week"
					If isFirst Then
						opr.Value = [date].AddDays(7)
					Else
						opr.Value = [date].AddDays(-7)
					End If
				Case "Month"
					If isFirst Then
						opr.Value = [date].AddMonths(1)
					Else
						opr.Value = [date].AddMonths(-1)
					End If
			End Select
		End Sub

		Private Sub InitData()
			dt = New DataTable()
			dt.Columns.Add("ID", GetType(Integer))
			dt.Columns.Add("Date", GetType(DateTime))
			Dim rnd As New Random()
			For i As Integer = 0 To 999
				dt.Rows.Add(New Object() { i, DateTime.Now.Date.AddDays(rnd.Next(100)) })
			Next i
			dv = New DataView(dt)
		End Sub

		Private dt As DataTable
		Private dv As DataView

		Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
			filterControl1.ApplyFilter()
			gridView1.ActiveFilterCriteria = filterControl1.FilterCriteria
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load

		End Sub
	End Class
End Namespace

File: CS/WindowsApplication197/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors.Repository;
using System.Reflection;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Filtering;
using DevExpress.Data.Filtering;
using DevExpress.Data.Filtering.Helpers;

namespace WindowsApplication197
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            InitData();
            gridControl1.DataSource = dv;

            DevExpress.XtraGrid.FilterEditor.ViewFilterColumnCollection columnCollection = new DevExpress.XtraGrid.FilterEditor.ViewFilterColumnCollection(gridView1);
            RepositoryItemPopupContainerEdit ri = new RepositoryItemPopupContainerEdit();
            ri.PopupControl = popupContainerControl1;
            ri.CloseOnOuterMouseClick = false;
            ri.QueryResultValue += new DevExpress.XtraEditors.Controls.QueryResultValueEventHandler(ri_QueryResultValue);
            ri.QueryPopUp += new CancelEventHandler(ri_QueryPopUp);
            columnCollection["Date"].SetColumnEditor(ri);
            columnCollection["ID"].SetColumnEditor(new RepositoryItemCalcEdit());
            filterControl1.SetFilterColumnsCollection(columnCollection);

        }

        void ri_QueryPopUp(object sender, CancelEventArgs e)
        {
            dateEdit1.EditValue = ((PopupContainerEdit)sender).EditValue;
        }

        void ri_QueryResultValue(object sender, DevExpress.XtraEditors.Controls.QueryResultValueEventArgs e)
        {
            e.Value = dateEdit1.EditValue;
            DateTime date = Convert.ToDateTime(e.Value);
            PropertyInfo pi = typeof(FilterControl).GetProperty("FocusInfo", BindingFlags.Instance| BindingFlags.NonPublic);
            FilterControlFocusInfo focus = (FilterControlFocusInfo)pi.GetValue(filterControl1, null);
            ClauseNode cnode = focus.Node as ClauseNode;
            if (cnode == null) return;
            if (cnode.Operation == ClauseType.Between)
            {
                if (focus.ElementIndex == 2)
                {
                    PatchOperator(cnode.AdditionalOperands[1] as OperandValue, date, comboBoxEdit1.Text, true);
                }
                else if (focus.ElementIndex == 3)
                {
                    PatchOperator(cnode.AdditionalOperands[0] as OperandValue, date, comboBoxEdit1.Text, false);
                }
                cnode.RecalcLabelInfo();
            }
        }

        void PatchOperator(OperandValue opr, DateTime date, string range, bool isFirst)
        {
            if (((object)opr) == null) return;
            switch (range)
            {
                case "Day": opr.Value = date.AddDays(isFirst ? 1 : -1); break;
                case "Week": opr.Value = date.AddDays(isFirst ? 7 : -7); break;
                case "Month": opr.Value = date.AddMonths(isFirst ? 1 : -1); break;
            }
        }

        private void InitData()
        {
            dt = new DataTable();
            dt.Columns.Add("ID", typeof(int));
            dt.Columns.Add("Date", typeof(DateTime));
            Random rnd= new Random();
            for (int i = 0; i < 1000; i++)
            {
                dt.Rows.Add(new object[] { i, DateTime.Now.Date.AddDays(rnd.Next(100)) });
            }
            dv = new DataView(dt);
        }

        DataTable dt;
        DataView dv;

        private void simpleButton1_Click(object sender, EventArgs e)
        {
            filterControl1.ApplyFilter();
            gridView1.ActiveFilterCriteria = filterControl1.FilterCriteria;
        }

        private void Form1_Load(object sender, EventArgs e)
        {

        }
    }
}